plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java

```diff
@@ -0,0 +1,180 @@
+import org.opensearch.plugin.insights.rules.model.MetricType;
+import org.opensearch.plugin.insights.rules.model.SearchQueryRecord;
+import org.opensearch.plugin.insights.settings.QueryInsightsSettings;
+import org.opensearch.threadpool.Scheduler;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.LinkedBlockingQueue;
+
+/**
+ * Service responsible for gathering, analyzing, storing and exporting
+ * information related to search queries
+ *
+ * @opensearch.internal
+ */
+public class QueryInsightsService extends AbstractLifecycleComponent {
+    /**
+     * The internal OpenSearch thread pool that execute async processing and exporting tasks
+     */
+    private final ThreadPool threadPool;
+
+    /**
+     * Services to capture top n queries for different metric types
+     */
+    private final Map<MetricType, TopQueriesService> topQueriesServices;
+
+    /**
+     * Flags for enabling insight data collection for different metric types
+     */
+    private final Map<MetricType, Boolean> enableCollect;
+
+    /**
+     * The internal thread-safe queue to ingest the search query data and subsequently forward to processors
+     */
+    private final LinkedBlockingQueue<SearchQueryRecord> queryRecordsQueue;
+
+    /**
+     * Holds a reference to delayed operation {@link Scheduler.Cancellable} so it can be cancelled when
+     * the service closed concurrently.
+     */
+    protected volatile Scheduler.Cancellable scheduledFuture;
+
+    /**
+     * Constructor of the QueryInsightsService
+     *
+     * @param threadPool     The OpenSearch thread pool to run async tasks
+     */
+    @Inject
+    public QueryInsightsService(final ThreadPool threadPool) {
+        enableCollect = new HashMap<>();
+        queryRecordsQueue = new LinkedBlockingQueue<>(QueryInsightsSettings.QUERY_RECORD_QUEUE_CAPACITY);
+        topQueriesServices = new HashMap<>();
+        for (MetricType metricType : MetricType.allMetricTypes()) {
+            enableCollect.put(metricType, false);
+            topQueriesServices.put(metricType, new TopQueriesService(metricType));
+        }
+        this.threadPool = threadPool;
+    }
+
+    /**
+     * Ingest the query data into in-memory stores
+     *
+     * @param record the record to ingest
+     */
+    public boolean addRecord(final SearchQueryRecord record) {
+        boolean shouldAdd = false;
+        for (Map.Entry<MetricType, TopQueriesService> entry : topQueriesServices.entrySet()) {
+            if (!enableCollect.get(entry.getKey())) {
+                continue;
+            }
+            List<SearchQueryRecord> currentSnapshot = entry.getValue().getTopQueriesCurrentSnapshot();
+            // skip add to top N queries store if the incoming record is smaller than the Nth record
+            if (currentSnapshot.size() < entry.getValue().getTopNSize()
```

Comment:
I think we need to incorporate the requirement where we want to record the queries with latency only above certain threshold. Here, any request would make to the queue even if its < 10ms, which are most of the times useless. I posted similar comment earlier somewhere in this PR. We can take it up as a separate PR later if it makes sense.

Reply:
Actually, with the most recent change, the request with latency "higher than the Nth top queries" will be added to the queue. As you mentioned before, we can also improve this even better by adding a lower bound config for requests to be collected.